### ISSUE TITLE
Issue in Shutdown timeout

### ISSUE THREAD

**Issue by rameezrajas** [spring-cloud/spring-cloud-stream-binder-rabbit#128]
Hi, 
We use Spring cloud-stream-binder-rabbit. Also, we configure rabbit queues using  <int-amqp:inbound-channel-adapter>. We want to configure a shutdown timeout for the inflight messages in the inbound adapter so that when i shutdown the service, it wait for the configured time. It seems this shutdown timeout is not working if we have spring cloud stream binder rabbit in our project.

if we dont have stream binder rabbit in my project, shutdown time out works as expected

When i investigated this, i found RabbitMessageChannelBinder.destroy method is shutting down the AMQP connection first and doShutdown method in SimpleMessageListenerContainer is not waiting for the configured shutdown timeout since the AMQP connection has been already shutdown by the destroy method. 

Is this how its supposed to work? What can i do so that doShutdown in SimpleMessageListenerContainer gets called first and the RabbitMessageChannelBinder.destroy gets called last.

**Comment by garyrussell** [comment]
>Is this how its supposed to work?

If that's actually what's happening then, no, that's not how it's supposed to work; the binder shouldn't be destroyed until all the `Lifecycle` beans are stopped.

**Comment by rameezrajas** [comment]
@garyrussell . Yes. Thats is what happening. Binder is getting destroyed before SimpleMessageListenerContainer are stopped. Is there any temporary fix i can do to get rid of this problem? A kind of implementing my own SmartLifeCycle bean to stop the SimpleMessageListenerContainer beans first?

**Comment by garyrussell** [comment]
It's a bug in the binder, introduced by this https://github.com/spring-cloud/spring-cloud-stream-binder-rabbit/pull/96 

The connection factory is destroyed before the bindings are stopped (via `InputBindingLifecycle`). We shouldn't be destroying the connection factory `@Bean` here.

cc: @artembilan 

I can't think of a workaround right now - 2.0 will have an actuator that allows stopping bindings but that won't help for 1.3.x.

**Comment by garyrussell** [comment]
Removing that destroy fixes it

```
2018-02-22 11:58:34.227  INFO 27082 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 2147483647
foo
2018-02-22 11:58:34.265  INFO 27082 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2018-02-22 11:58:34.270  INFO 27082 --- [           main] com.example.Scbr128Application           : Started Scbr128Application in 3.026 seconds (JVM running for 3.464)
2018-02-22 11:58:34.270  INFO 27082 --- [           main] ationConfigEmbeddedWebApplicationContext : Closing org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@691a7f8f: startup date [Thu Feb 22 11:58:31 EST 2018]; root of context hierarchy
2018-02-22 11:58:34.271  INFO 27082 --- [           main] s.c.a.AnnotationConfigApplicationContext : Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@52f57666: startup date [Thu Feb 22 11:58:34 EST 2018]; parent: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@691a7f8f
2018-02-22 11:58:34.272  INFO 27082 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Stopping beans in phase 2147483647
2018-02-22 11:58:34.273  INFO 27082 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Stopping beans in phase 2147482647
2018-02-22 11:58:34.274  INFO 27082 --- [           main] o.s.a.r.l.SimpleMessageListenerContainer : Waiting for workers to finish.
Listener Exiting
2018-02-22 11:58:37.238  INFO 27082 --- [           main] o.s.a.r.l.SimpleMessageListenerContainer : Successfully waited for workers to finish.
```

**Comment by garyrussell** [comment]
@rameezrajas Here is a workaround - replace boot's connection factory with this one...

```java
@Bean
public CachingConnectionFactory rabbitConnectionFactory(RabbitProperties config)
		throws Exception {
	RabbitConnectionFactoryBean factory = new RabbitConnectionFactoryBean();
	if (config.determineHost() != null) {
		factory.setHost(config.determineHost());
	}
	factory.setPort(config.determinePort());
	if (config.determineUsername() != null) {
		factory.setUsername(config.determineUsername());
	}
	if (config.determinePassword() != null) {
		factory.setPassword(config.determinePassword());
	}
	if (config.determineVirtualHost() != null) {
		factory.setVirtualHost(config.determineVirtualHost());
	}
	if (config.getRequestedHeartbeat() != null) {
		factory.setRequestedHeartbeat(config.getRequestedHeartbeat());
	}
	RabbitProperties.Ssl ssl = config.getSsl();
	if (ssl.isEnabled()) {
		factory.setUseSSL(true);
		if (ssl.getAlgorithm() != null) {
			factory.setSslAlgorithm(ssl.getAlgorithm());
		}
		factory.setKeyStore(ssl.getKeyStore());
		factory.setKeyStorePassphrase(ssl.getKeyStorePassword());
		factory.setTrustStore(ssl.getTrustStore());
		factory.setTrustStorePassphrase(ssl.getTrustStorePassword());
	}
	if (config.getConnectionTimeout() != null) {
		factory.setConnectionTimeout(config.getConnectionTimeout());
	}
	factory.afterPropertiesSet();
	CachingConnectionFactory connectionFactory = new CachingConnectionFactory(factory.getObject()) {

		private final AtomicBoolean ignoreFirstReset = new AtomicBoolean(true);

		@Override
		public void resetConnection() {
			if (this.ignoreFirstReset.getAndSet(false)) {
				return;
			}
			super.resetConnection();
		}

	};
	connectionFactory.setAddresses(config.determineAddresses());
	connectionFactory.setPublisherConfirms(config.isPublisherConfirms());
	connectionFactory.setPublisherReturns(config.isPublisherReturns());
	if (config.getCache().getChannel().getSize() != null) {
		connectionFactory
				.setChannelCacheSize(config.getCache().getChannel().getSize());
	}
	if (config.getCache().getConnection().getMode() != null) {
		connectionFactory
				.setCacheMode(config.getCache().getConnection().getMode());
	}
	if (config.getCache().getConnection().getSize() != null) {
		connectionFactory.setConnectionCacheSize(
				config.getCache().getConnection().getSize());
	}
	if (config.getCache().getChannel().getCheckoutTimeout() != null) {
		connectionFactory.setChannelCheckoutTimeout(
				config.getCache().getChannel().getCheckoutTimeout());
	}
	return connectionFactory;
}
```

It ignores the first call to `resetConnection()`.

**Comment by garyrussell** [comment]
@rameezrajas An alternative work around would be to wrap the factory in a proxy...

```java
@Bean
public static BeanPostProcessor ccfBpp() {
	return new BeanPostProcessor() {

		private final AtomicBoolean ignoreFirstReset = new AtomicBoolean(true);

		@Override
		public Object postProcessBeforeInitialization(Object bean, String name) throws BeansException {
			return bean;
		}

		@Override
		public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
			if (bean instanceof CachingConnectionFactory) {
				MethodInterceptor interceptor = new MethodInterceptor() {

					@Override
					public Object invoke(MethodInvocation invocation) throws Throwable {
						if (invocation.getMethod().getName().equals("destroy")
								&& ignoreFirstReset.getAndSet(false)) {
							return null;
						}
						return invocation.proceed();
					}

				};
				if (AopUtils.isAopProxy(bean)) {
					((Advised) bean).addAdvice(interceptor);
					return bean;
				}
				else {
					ProxyFactory proxyFactory = new ProxyFactory(bean);
					proxyFactory.addAdvice(interceptor);
					return proxyFactory.getProxy();
				}
			}
			else {
				return bean;
			}
		}

	};
}
```

**Comment by rameezrajas** [comment]
@garyrussell Thank you for the quick turnaround. Your second workaround to wrap the factory in a proxy seems to be working for me. You first workaround did not work for me. We can live this until 2.0 is released. Thanks again.

**Comment by garyrussell** [comment]
Great - not sure why the first wouldn't work; it's essentially doing the same thing; not important though, I guess.

The fix will be in the 1.3.4.RELEASE too.
